src/lsm/schema.zig

```diff
@@ -334,6 +335,65 @@ pub const TableData = struct {
     }
 };
 
+pub const FreeSetNode = struct {
+    pub const Word = u64;
+
+    pub const Metadata = extern struct {
+        next_free_set_block_checksum: u128,
+        next_free_set_block_address: u64,
+        reserved: [104]u8 = .{0} ** 104,
+
+        comptime {
+            assert(stdx.no_padding(Metadata));
+            assert(@sizeOf(Metadata) == vsr.Header.Block.metadata_size);
+        }
+    };
+
+    fn metadata(free_set_block: BlockPtrConst) *const Metadata {
+        const header = header_from_block(free_set_block);
+        assert(header.command == .block);
+        assert(header.block_type == .free_set);
+        assert(header.address > 0);
+        assert(header.snapshot == 0);
+
+        const header_metadata = std.mem.bytesAsValue(Metadata, &header.metadata_bytes);
+        assert(stdx.zeroed(&header_metadata.reserved));
+
+        if (header_metadata.next_free_set_block_address == 0) {
+            assert(header_metadata.next_free_set_block_checksum == 0);
+        }
+
+        assert((header.size - @sizeOf(vsr.Header)) % @sizeOf(Word) == 0);
+
+        return header_metadata;
+    }
+
+    pub fn assert_valid_header(free_set_block: BlockPtrConst) void {
+        _ = metadata(free_set_block);
+    }
+
+    pub fn next(free_set_block: BlockPtrConst) ?struct { checksum: u128, address: u64 } {
```

Comment:
I think we should add `vsr.BlockReference` as it is used in a couple of different places. But that's for a follow up.